keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -233,7 +233,606 @@ yet but we think may be interesting to consider in the future.
+promptness.
+
+For implementation simplicity (see below), let us use the same initial
+service time guess for every request; call that duration G.  A good
+choice might be the service time limit (1 minute).  Different guesses
+will give slightly different dynamics, but any positive number can be
+used for G without ruining the long-term behavior.
+
+As in ordinary fair queuing, there is a bound on divergence from the
+ideal.  In plain fair queuing the bound is one packet; in our version
+it is C read-only requests.
+
+To support efficiently making the necessary adjustments once a
+request’s actual service time is known, the virtual finish time of a
+request and the last virtual finish time of a queue are not
+represented directly but instead computed from queue length, request
+position in the queue, and an alternate state variable that holds the
+queue’s virtual start time.  While the queue is empty and has no
+requests executing: the value of its virtual start time variable is
+ignored and its last virtual finish time is considered to be in the
+virtual past.  When a request arrives to an empty queue with no
+requests executing, the queue’s virtual start time is set to `now()`.
+The virtual finish time of request number J in the queue (counting
+from J=1 for the head) is J * G + (virtual start time).  While the
+queue is non-empty: the last virtual finish time of the queue is the
+virtual finish time of the last request in the queue.  While the queue
+is empty and has a request executing: the last virtual finish time is
+the queue’s virtual start time.  When a request is dequeued for
+service the queue’s virtual start time is advanced by G.  When a
+request finishes being served, and the actual service time was S, the
+queue’s virtual start time is decremented by G - S.
+
+### Example Configuration
+
+
+For requests from admins and requests in service of other, potentially
+system, requests.
+```yaml
+kind: RequestPriority
+meta:
+  name: system-top
+spec:
+  priorityLevel: 0
+```
+
+For system self-maintenance requests.
+```yaml
+kind: RequestPriority
+meta:
+  name: system-high
+spec:
+  priorityLevel: 1000
+  assuredConcurrencyShares: 10
+  queuesPerWidth: 128
+  handSize: 6
+  queueLengthLimit: 100
+```
+
+For the garbage collector.
+```yaml
+kind: RequestPriority
+meta:
+  name: system-low
+spec:
+  priorityLevel: 2000
+  assuredConcurrencyShares: 10
+  queuesPerWidth: 1
+  queueLengthLimit: 1000
+```
+
+For user requests from kubectl.
+```yaml
+kind: RequestPriority
+meta:
+  name: workload-high
+spec:
+  priorityLevel: 9000
+  assuredConcurrencyShares: 10
+  queuesPerWidth: 128
+  handSize: 6
+  queueLengthLimit: 100
+```
+
+For requests from controllers processing workload.
+```yaml
+kind: RequestPriority
+meta:
+  name: workload-low
+spec:
+  priorityLevel: 10000
+  assuredConcurrencyShares: 10
+  queuesPerWidth: 128
+  handSize: 6
+  queueLengthLimit: 100
+```
+
+Some flow schemata.
```

Comment:
I believe the api model of flow schema here can be simplified by:
- removing support for `patternMatch` and `nonPatternMatch`. additionally, all controllers are sharing the same user identity as `system:kube-controller-manager` but differs in user-agent http header.
- make the field `matchingPriority` optional and only filled to resolve conflicts among multiple flow schemas. A default value will be working for most cases I suppose.
- instead of defining explicit boolean tests in the api model. consider make it a bit simpler by costing a bit ambiguity. as the yaml examples in the following shows:

```yaml
kind: FlowSchema
meta:
  name: system-top
spec:
  matchingPriority: 500 // default to 500, override if necessary
  requestPriority:
    name: system-top
  matchRules:
  - field: groups
     valueSet: [ "system:masters" ]
---
kind: FlowSchema
meta:
  name: system-high-node-heartbeat
spec:
  matchingPriority: 500 // default to 500, override if necessary
  requestPriority:
    name: system-high
  flowDistinguisher:
    source: user
    # no transformation in this case
    matchRules:
  - field: groups
     valueSet: [ "system:nodes" ]
  - field: resource
     valueSet: [ "nodes" ]
---
kind: FlowSchema
meta:
  name: system-high-system-objects
spec:
  matchingPriority: 500 // default to 500, override if necessary
  requestPriority:
    name: system-high
  flowDistinguisher:
    source: user
    # no transformation in this case
    matchRules:
  - field: groups
     valueSet: [ "system:nodes" ]
  - field: namespace
     valueSet: [ "kube-system" ]
---
kind: FlowSchema
meta:
  name: system-high-system-objects
spec:
  matchingPriority: 500 // default to 500, override if necessary
  requestPriority:
    name: system-high
  flowDistinguisher:
    source: user
    # no transformation in this case
    matchRules:
  - field: groups
     valueSet: [ "system:nodes" ]
  - field: namespace
     valueSet: [ "kube-system" ]
---
kind: FlowSchema
meta:
  name: workload-low-serviceaccounts
spec:
  matchingPriority: 500 // default to 500, override if necessary
  requestPriority:
    name: workload-high
  flowDistinguisher:
    source: namespace
    # no transformation in this case
  matchRules:
  - field: groups
     valueSet: [ "system:serviceaccounts" ]
....
```

Reply:
> * all controllers are sharing the same user identity as `system:kube-controller-manager` but differs in user-agent http header.

that's not generally accurate... typically, each controller loop gets a distinct identity